### PR TITLE
Drone Fails if New Changes Are Made to Files while running tests

### PR DESCRIPTION
**Pull request by dmcavoy**
Raises an error in the run of unit tests and UI tests if changes to `schema.rb` or `dashboard/config/locales/*` are generated during while running the tests on drone. 

Checkout the Drone run on c621dd1 where the error is raised because of the change made in 96f0f2d. Images below are from that run:

Unit Tests:
Screen Shot 2020-09-10 at 8 02 08 PM

UI Tests:
Screen Shot 2020-09-10 at 8 02 43 PM

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLAT-255)

## Testing story

See above showing the testing I did with commits in this branch. Not sure how else to test?

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: lib/rake/circle.rake [hunk 1]
```diff
@@ -71,6 +71,13 @@ namespace :circle do
     else
       RakeUtils.rake_stream_output 'test:changed'
     end
+
+    if GitUtils.changed_in_branch_or_local?(GitUtils.current_branch, ['dashboard/config/locales/*.en.yml'])
+      raise 'Unexpected change to dashboard/config/locales/ - Make sure you running seeding locally and include those changes in your branch.'
```
Comment (uponthesun) [inline lib/rake/circle.rake]: Nit - "you running"

**Review by uponthesun — COMMENTED** [review]
🎉 

I'd recommend extracting the checks into a function, so they stay in sync in case we need to change them at some point. You should be able to just throw it at the bottom of the file.

File: lib/rake/circle.rake [hunk 2]
```diff
@@ -128,6 +135,13 @@ namespace :circle do
     end
     close_sauce_connect if use_saucelabs || test_eyes?
     RakeUtils.system_stream_output 'sleep 10'
+
+    if GitUtils.changed_in_branch_or_local?(GitUtils.current_branch, ['dashboard/config/locales/*.en.yml'])
+      raise 'Unexpected change to dashboard/config/locales/ - Make sure you running seeding locally and include those changes in your branch.'
+    end
+    if GitUtils.changed_in_branch_or_local?(GitUtils.current_branch, ['dashboard/db/schema.rb'])
+      raise 'Unexpected change to schema.rb - Make sure you run your migration locally and push those changes into your branch.'
+    end
```
Comment (davidsbailey) [inline lib/rake/circle.rake]: nice helpful error messages, Dani! I was going to suggest turning this into an array people can add to, but I like how this gives a more useful error message in each case instead.

my one concern is that I might not know what to do next if I run migrate or seed and the problem does not go away. To address this, I would recommend including a diff of the affected file or directory, like we do here: https://github.com/code-dot-org/code-dot-org/blob/daaec080bb05d53e1292ade5708135fe16527249/lib/rake/circle.rake#L137-L141

for example, you should be able to add a line like one of these before you `raise`: 
```
RakeUtils.system_stream_output('git diff -- dashboard/db/schema.rb | cat')
RakeUtils.system_stream_output('git diff -- dashboard/config/locales | cat')
```

**Review by davidsbailey — COMMENTED** [review]
Awesome work Dani! One request inline.

**Comment by dmcavoy** [comment]
@davidsbailey @uponthesun Is it possible the after reverting the things that would cause the change to scripts.en.yml that that change might still be around in the branch? Do I need to add something to clear it out?

**Comment by davidsbailey** [comment]
> @davidsbailey @uponthesun Is it possible the after reverting the things that would cause the change to scripts.en.yml that that change might still be around in the branch? Do I need to add something to clear it out?

I don't think running seed would ever remove anything from that file. So yes, you might need to clear out changes to that file manually if you previously ran seed and committed the additions.

**Comment by dmcavoy** [comment]
@uponthesun @davidsbailey It looks like my test (1643918) did fail but only in the UI tests and not the unit tests. Does that make sense?

**Comment by davidsbailey** [comment]
> @uponthesun @davidsbailey It looks like my test ([1643918](https://github.com/code-dot-org/code-dot-org/commit/1643918a869658a25d3ed88e52038f0ec74a11b2)) did fail but only in the UI tests and not the unit tests. Does that make sense?

Yes, I think it does, because only the UI test container runs `rake seed`, which is what generates the scripts.en.yml changes. the unit test container uses `rake db:test:prepare` to initialize the DB only with test fixture data.

I'm not exactly sure where I would expect schema.rb changes to fail. @uponthesun do you know if we run migrations in both Drone containers? or do we just initialize from schema.rb itself? if the latter... than we may not be catching problems with schema.rb. I went to check the output of a drone run just now, but the drone links are not working for me at the moment.

**Comment by dmcavoy** [comment]
@davidsbailey I only tested the scripts.en.yml because I was trying to avoid making a migration and then reverting it but i'm not realizing I wouldn't need to run it because thats the whole point of this PR. Anyway all of that to say I probably still need to test that

**Comment by dmcavoy** [comment]
Looks like the Unit tests picked up the unrun migration in 24c93b8 @davidsbailey

**Comment by davidsbailey** [comment]
> Looks like the Unit tests picked up the unrun migration in [24c93b8](https://github.com/code-dot-org/code-dot-org/commit/24c93b83791f95a2816699963dad2b990c6d778e) @davidsbailey

Woohoo, excellent! From reading the drone scripts, it looks like they do rake install and then rake build. So I would guess that this works because we initialize the database from schema.rb during rake install, and then run migrations during rake build.

Thank you for verifying this. I think this will eliminate a lot of headaches!

**Review by davidsbailey — APPROVED** [review]
Looks good! Thanks for your persistence on this, Dani!

**Pull request merged**